Make new address form that uses SchemaForm

erikphansen opened the pull request:
## Description
This PR creates a new `AddressFormV2` component and uses it in place of the `AddressForm` component when running the site locally.

Known issues so far:
- [x] When selecting the "I live on a military base..." checkbox, the form should automatically select United States as the country, if it isn't already.
- [x] When selecting the `My home address is the same as my mailing address.` checkbox on the Home Address edit modal, it crashes the React app :( Obviously this is a severe bug I need to address ASAP.
- [x] If a Vet has an overseas military base address on file, when the edit modal opens the "I live on a military base..." checkbox should be pre-checked.
- [x] Disable and gray out the country dropdown if "I live on a military base..." is checked. This requires a change to the form system, which is handled in this PR: https://github.com/department-of-veterans-affairs/vets-website/pull/11617
- [x] Form could crash on load if initial data values were null, which happens all the time with the real service. Sadly our local mock API was not using null values so I was late in discovering this.

## Testing done
Lots of local testing + some new unit tests

## Screenshots
**GIF showing the form UI changing depending on its data:**
![ScreenFlow](https://user-images.githubusercontent.com/20728956/73670802-be316600-465e-11ea-8d09-41b50ef967ca.gif)

**GIF showing form submission works as usual**
![address-2](https://user-images.githubusercontent.com/20728956/73411363-f715b780-42b9-11ea-8840-76d5a73476f5.gif)

**GIF showing the country dropdown switching to USA and getting disabled**
![usa-disabled](https://user-images.githubusercontent.com/20728956/73684179-846d5900-4678-11ea-978d-281fce232851.gif)

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

erikphansen committed:
Make new address form that uses SchemaForm

WIP: making new address schema form

Mostly cleanup + one function rename

Update schema when livesOnMilitaryBase

Update the options for country, city, and state when the "lives on military base" checkbox is selected.

update country when 'lives on base' is checked

Add form system option to not delete hidden field data

Stop cleaning data on each data change

to stop deleting data for fields that are no longer being shown. We want to persist that data in case the user wants to see those fields again. We will clean up on submission instead.

Refactor DOM of Vet360EditModal

we don't want to wrap the SchemaForm in a form since the SchemaForm renders its own form element

fix failing unit test

Only show checkbox when we should

erikphansen commented on src/platform/forms-system/src/js/state/helpers.js:
```diff
@@ -478,14 +478,21 @@ export function updateItemsSchema(schema, fieldData = null) {
  * This is the main sequence of updates that happens when data is changed
- * on a form. Most updates are applied to the schema, except that the data
- * is updated to remove newly hidden data
+ * on a form. Most updates are applied to the schema. And by default the data
```
All the changes in this file are also part of this PR: https://github.com/department-of-veterans-affairs/vets-website/pull/11589

erikphansen commented on src/platform/user/profile/vet360/components/AddressField/AddressEditModal.jsx:
```diff
@@ -7,6 +7,11 @@ import Vet360EditModal from '../base/Vet360EditModal';
+
+import environment from 'platform/utilities/environment';
+
+const useNewAddressForm = environment.isLocalhost();
```
Right now we only want to use this new address form locally as we are about to start manual QA on a different feature that uses the existing form and we don't want to create failures during QA that have nothing to do with the feature they are really trying to test.

erikphansen commented on src/platform/user/profile/vet360/components/AddressField/AddressField.jsx:
```diff
@@ -114,8 +135,9 @@ export const validateCleanData = (
-export const convertCleanDataToPayload = (cleanData, fieldName) =>
-  pickBy(
+export const convertCleanDataToPayload = (data, fieldName) => {
+  const cleanData = convertNextValueToCleanData(data);
```
We need to call the `convertNextValueToCleanData()` here as part of the pre-submission transformation to work with the new address form. This change will also affect staging and prod but will not have any side-effects as this helper is already called every single time that the user changes data on the current address form.

ncksllvn replied:
Oh how I wish we could get rid of that function lol.

erikphansen replied:
It's basically functioning like a `transformForSubmit` function now which is nice :)

erikphansen commented on src/platform/user/profile/vet360/components/base/Vet360EditModal.jsx:
```diff
@@ -77,13 +90,52 @@ export default class Vet360EditModal extends React.Component {
     const isLoading = transactionRequest && transactionRequest.isPending;
     const error = transactionRequest && transactionRequest.error;
 
+    const actionButtons = (
```
Had to do a fair bit of refactoring in this file since with the new form:
1. We can't wrap the new AddressFormV2 in a `form` element because it renders its own `form` element and a `form` in a `form` isn't valid
2. We need to pass the `Vet360EditModalActionButtons` in as a child element of the new AddressFormV2

erikphansen committed:
Fix copy mailing address functionality

erikphansen committed:
remove unused attribute

erikphansen commented on src/platform/user/profile/vet360/components/AddressField/AddressField.jsx:
```diff
@@ -135,6 +157,215 @@ export const convertCleanDataToPayload = (cleanData, fieldName) =>
+    'ui:description': () => (
+      <div className="vads-u-padding-x--2p5">
+        <AdditionalInfo
+          lassName="vads-u-margin-left--0"
```
I'm removing this line completely. The original address form that we're replacing has this same typo. It looks fine without it working. And the `AdditionalInfo` design system component doesn't seem to support a `className` prop anyway.

ncksllvn reviewed (approved):
Thanks for doing this!

erikphansen committed:
refactor MILITARY_STATES creation

erikphansen committed:
Force country to USA when 'livesOnBase' is checked

Create a simple wrapper function to transform the form data before calling the Redux action creator.

erikphansen committed:
Optionally pre-check the `livesOnBase` checkbox

When popping open the edit mailing address modal, if the current address is on a military base, the corresponding checkbox should be pre-selected.

erikphansen committed:
Gray out disabled select fields

erikphansen committed:
disable country dropdown when needed

erikphansen commented on src/platform/forms-system/src/js/state/helpers.js:
```diff
@@ -349,6 +349,53 @@ export function updateSchemaFromUiSchema(
   return currentSchema;
 }
 
+/**
```
These changes are also in a smaller PR here https://github.com/department-of-veterans-affairs/vets-website/pull/11617

erikphansen commented on src/platform/forms-system/src/js/widgets/SelectWidget.jsx:
```diff
@@ -34,7 +34,9 @@ function SelectWidget({
       name={id}
       multiple={multiple}
-      className={options.widgetClassNames}
+      className={`${options.widgetClassNames} ${
```
enabled vs disabled with this change:
![image](https://user-images.githubusercontent.com/20728956/73681734-d8297380-4673-11ea-8f5e-f82e7c85d526.png)
This new look is already being done on the existing profile address update form

cvalarida replied:
For what it's worth, we've intentionally shied away from the "disabled" pattern in the past. :man_shrugging:

erikphansen commented on src/platform/forms-system/test/js/state/helpers.unit.spec.js:
```diff
@@ -1,10 +1,12 @@
 import { expect } from 'chai';
+import sinon from 'sinon';
```
All the changes in this file are also done as a smaller PR here https://github.com/department-of-veterans-affairs/vets-website/pull/11617

erikphansen commented on src/platform/user/profile/vet360/components/AddressField/AddressField.jsx:
```diff
@@ -135,6 +157,224 @@ export const convertCleanDataToPayload = (cleanData, fieldName) =>
+          enum: ADDRESS_FORM_VALUES.COUNTRIES,
+        };
+      },
+      updateUiSchema: formData => {
```
Support for this new `udpateUiSchema` hook is added here 
![image](https://user-images.githubusercontent.com/20728956/73681899-1a52b500-4674-11ea-838d-435b6d0787b8.png)

erikphansen commented on src/platform/user/profile/vet360/util/local-vet360.js:
```diff
@@ -79,7 +79,7 @@ export const mockContactInformation = {
     addressLine3: 'string',
     addressPou: 'CORRESPONDENCE',
-    addressType: 'domestic',
+    addressType: 'DOMESTIC',
```
this now matches reality: https://github.com/department-of-veterans-affairs/vets-website/blob/0f539058825668a08f9518f56c85dec9dd96d852/src/platform/user/profile/vet360/constants/index.js#L12-L16

erikphansen commented on src/platform/user/profile/vet360/util/local-vet360.js:
```diff
@@ -102,7 +102,7 @@ export const mockContactInformation = {
     addressLine3: '',
     addressPou: 'RESIDENCE/CHOICE',
-    addressType: 'MILITARY OVERSEAS',
+    addressType: 'OVERSEAS MILITARY',
```
this now matches reality: https://github.com/department-of-veterans-affairs/vets-website/blob/0f539058825668a08f9518f56c85dec9dd96d852/src/platform/user/profile/vet360/constants/index.js#L12-L16

erikphansen committed:
refactor ui unit test

cvalarida commented on src/applications/personalization/profile360/sass/profile-360.scss:
```diff
@@ -1,4 +1,5 @@
 @import "~@department-of-veterans-affairs/formation/sass/shared-variables";
+@import "../../../../platform/forms/sass/m-schemaform";
```
:grimacing: Any idea how much this adds to your CSS bundle?

cvalarida commented on src/platform/user/profile/vet360/components/AddressField/AddressField.jsx:
```diff
@@ -135,6 +157,224 @@ export const convertCleanDataToPayload = (cleanData, fieldName) =>
   );
+};
+
+const formSchema = {
```
Would it be useful to share this schema with the back end?

cvalarida commented on src/platform/user/profile/vet360/components/AddressField/AddressField.jsx:
```diff
@@ -135,6 +157,224 @@ export const convertCleanDataToPayload = (cleanData, fieldName) =>
+  required: ['countryName', 'addressLine1', 'city'],
+};
+
+const uiSchema = {
```
I might recommend pulling these out into a separate file just because of the added length.

cvalarida reviewed (approved):
Interesting stuff! I mostly only looked at the stuff that uses the schema / uiSchema since I don't have context for the rest of it myself, but it looks good. :+1:

I'd be really interested to hear how you're using bits of the forms system without using the whole thing some time!

erikphansen committed:
move schemas, etc. to their own file

erikphansen committed:
Trigger Jenkins

erikphansen committed:
Further cleaning of initial form data

String fields were getting set to empty strings which led to form validation errors or worse if they were set to null initially the app would crash due to null not being a valid value for the string field.

erikphansen committed:
Trigger Jenkins

erikphansen committed:
fix e2e test
